### PR TITLE
Remove arbitrary number of elements from QueueFile

### PR DESCRIPTION
**Pull request by f2prateek**

### CODE REVIEW

File: tape/src/main/java/com/squareup/tape/FileObjectQueue.java [hunk 1]
```diff
@@ -102,6 +101,18 @@ public FileObjectQueue(File file, Converter<T> converter) throws IOException {
     }
   }
 
+  public final void remove(int n) throws IOException {
+    try {
+      queueFile.remove(n);
+      if (listener != null) {
+        for (int i = 0; i < n; i++)
```
Comment (JakeWharton) [inline tape/src/main/java/com/squareup/tape/FileObjectQueue.java]: braces for body

Comment (f2prateek) [inline tape/src/main/java/com/squareup/tape/FileObjectQueue.java]: Ack, surprised both me and checkstyle and missed that.

File: tape/src/main/java/com/squareup/tape/QueueFile.java [hunk 2]
```diff
@@ -542,6 +542,41 @@ public synchronized void remove() throws IOException {
...
+  /**
+   * Removes the eldest {@code n} elements.
+   *
+   * @throws java.util.NoSuchElementException if the queue is empty
+   */
+  public synchronized void remove(int n) throws IOException {
+    if (isEmpty()) throw new NoSuchElementException();
+    if (n == 1) {
+      remove();
+    } else if (n >= elementCount) {
```
Comment (JakeWharton) [inline tape/src/main/java/com/squareup/tape/QueueFile.java]: Including greater than here is interesting behavior. This means that a loop of 10 `remove(1)` vs. `remove(10)` won't have the same behavior. I think we should throw in the greater than case.

File: tape/src/main/java/com/squareup/tape/FileObjectQueue.java [hunk 3]
```diff
@@ -102,6 +101,19 @@ public FileObjectQueue(File file, Converter<T> converter) throws IOException {
...
+  public final void remove(int n) throws IOException {
+    try {
+      queueFile.remove(n);
+      if (listener != null) {
+        for (int i = 0; i < n; i++) {
+          listener.onRemove(this);
+        }
+      }
+    } catch (IOException e) {
+      throw new FileException("Failed to remove.", e, file);
```
Comment (JakeWharton) [inline tape/src/main/java/com/squareup/tape/FileObjectQueue.java]: This is the worst thing I've ever written in my life. It's not entirely my fault that it's in here, but I blame me for allowing it to remain when we yanked this crap out of the old, old Retrofit project.

File: tape/src/main/java/com/squareup/tape/QueueFile.java [hunk 4]
```diff
@@ -542,6 +542,41 @@ public synchronized void remove() throws IOException {
...
 
+  /**
+   * Removes the eldest {@code n} elements.
+   *
+   * @throws java.util.NoSuchElementException if the queue is empty
+   */
+  public synchronized void remove(int n) throws IOException {
+    if (isEmpty()) throw new NoSuchElementException();
+    if (n == 1) {
+      remove();
```
Comment (JakeWharton) [inline tape/src/main/java/com/squareup/tape/QueueFile.java]: I would prefer to invert these. Have `remove()` call `remove(1)`.

File: tape/src/main/java/com/squareup/tape/QueueFile.java [hunk 5]
```diff
@@ -519,26 +519,42 @@ public synchronized int size() {
...
+   * Removes the eldest {@code n} elements.
+   *
+   * @throws java.util.NoSuchElementException if the queue is empty
+   */
+  public synchronized void remove(int n) throws IOException {
     if (isEmpty()) throw new NoSuchElementException();
-    if (elementCount == 1) {
+    if (n > elementCount) {
+      throw new IllegalArgumentException(
+          "Cannot remove more elements (" + n + ") than present in queue (" + elementCount + ".");
```
Comment (JakeWharton) [inline tape/src/main/java/com/squareup/tape/QueueFile.java]: Missing a closing parenthesis in your message

File: tape/src/main/java/com/squareup/tape/QueueFile.java [hunk 6]
```diff
@@ -519,26 +519,42 @@ public synchronized int size() {
...
+    } else if (n == elementCount) {
```
Comment (JakeWharton) [inline tape/src/main/java/com/squareup/tape/QueueFile.java]: This is going to be stupid nitpicky, but I think this method reads better like this:

```
if (empty) throw;
if (n > count) {
  throw;
}
if (n == count) {
  clear();
  return;
}

normal multi-remove
```
This way there's no cognitive load with the conditionals. Two state checks, one optimization check, and then the normal method body. I never have to think about `else` or `elseif` and what was already checked.

File: tape/src/test/java/com/squareup/tape/QueueFileTest.java [hunk 7]
```diff
@@ -139,6 +139,26 @@
     }
   }
 
+  @Test public void removeMultipleDoesNotCorrupt() throws IOException {
```
Comment (JakeWharton) [inline tape/src/test/java/com/squareup/tape/QueueFileTest.java]: Strange method name, but it matches the other one... ¯\_(ツ)_/¯

**Comment by JakeWharton** [comment]
Two tiny things. Then LGTM!

**Pull request merged**